File: 44wildcardMatching/wildcardMatching.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        std::vector<bool> dp(std::size(p) + 1);
        dp.front() = true;
        for (std::size_t column{1}; column != std::size(dp); ++column)
            if (p.at(column - 1) == '*') dp.at(column) = dp.at(column - 1);
        auto prev{dp};
        for (std::size_t row{1}; row != std::size(s) + 1; ++row)
        {
            dp.front() = false;
            for (std::size_t column{1}; column != std::size(dp); ++column)
            {
                if (s.at(row - 1) == p.at(column - 1) || p.at(column - 1) == '?') dp.at(column) = prev.at(column - 1);
                else if (p.at(column - 1) == '*') dp.at(column) = dp.at(column - 1) || prev.at(column);
                else dp.at(column) = false;
            }
            std::swap(prev, dp);
        }
        return prev.back();
    }
};

File: 44wildcardMatching/wildcardMatching.js
/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
var isMatch = function(s, p) {
    s = [...s]
    p = [...p]
    const dp = [true, ...Array(p.length).fill(false)]
    for (let column = 1; !Object.is(column, dp.length); ++column) dp[column] = dp[column - 1] && Object.is(p[column - 1], '*')
    for (let row = 1; !Object.is(row, s.length + 1); ++row)
    {
        const prev = [...dp]
        dp[0] = false
        for (let column = 1; !Object.is(column , dp.length); ++column)
        {
            if (Object.is(s[row - 1], p[column - 1]) || Object.is(p[column - 1], '?')) dp[column] = prev[column - 1]
            else if (Object.is(p[column - 1], '*')) dp[column] = dp[column - 1] || prev[column]
            else dp[column] = false
        }
    }
    return dp[dp.length - 1]
};
